core/src/main/scala/com/mesosphere/usi/core/SchedulerAdapter.scala

```diff
@@ -0,0 +1,199 @@
+import com.mesosphere.usi.core.models.{SpecEvent, SpecUpdated, SpecsSnapshot, StateEvent, StateSnapshot}
+
+import scala.concurrent.{Future, Promise}
+
+
+/**
+  * SchedulerAdapter provides a set of simple interfaces to the scheduler flow.
+  * @param schedulerFlow
+  * @param mat
+  */
+class SchedulerAdapter(schedulerFlow: Flow[SpecInput, StateOutput, NotUsed])(implicit mat: Materializer) {
+
+  private val extractSnapshot: Flow[SpecEvent, (SpecsSnapshot, Source[SpecUpdated, Any]), NotUsed] = Flow[SpecEvent]
+    .prefixAndTail(1) // First event must be a snashot, let's take it
+    .map {
+      case (Seq(specsSnapshot: SpecsSnapshot), source) =>
+        val specUpdatedSource = source.map {
+          case specUpdated: SpecUpdated => specUpdated
+          case other => throw new IllegalArgumentException(s"expected SpecUpdated but got $other")
+        }
+        (specsSnapshot, specUpdatedSource)
+      case _ =>
+        throw new IllegalArgumentException("scheduler spec stream must start with a snapshot")
+    }
+
+  private val mergeSnapshot: Flow[(StateSnapshot, Source[StateEvent, Any]), StateEvent, NotUsed] = Flow[StateOutput]
+    .flatMapConcat {
+      case (snapshot, updates) =>
+        updates.prepend(Source.single(snapshot))
+    }
+
+
+  /**
+    * Represents the scheduler as a Sink and Source.
+    *
+    * This method will materialize the scheduler first, then Sink and Source can be materialized independently.
+    * Please note that the first spec event coming to the Sink must be a SpecsSnapshot otherwise the stream will fail.
+    *
+    * @return Source which produces StateEvents and Sink which accepts SpecEvents
+    */
+  def asSourceAndSink(): (Source[StateEvent, NotUsed], Sink[SpecEvent, NotUsed]) = {
+
+    val (stateQueue, stateSource) = Source.queue[StateEvent](1, OverflowStrategy.backpressure).preMaterialize()
+
+    val (specQueue, specSink) = Sink.queue[SpecEvent]().preMaterialize()
+
+
+
+    Source.maybe
+      .prepend {
+        Source.unfoldResourceAsync[SpecEvent, SinkQueueWithCancel[SpecEvent]](
+          create = () => Future.successful(specQueue),
+          read = q => q.pull(),
+          close = q => Future.successful {
+            q.cancel()
+            Done
+          })
+      }
+      .via(extractSnapshot)
+      .via(schedulerFlow)
+      .via(mergeSnapshot)
+      .mapAsync(1)(stateQueue.offer)
+      .runWith(Sink.ignore)
+
+    (stateSource, specSink)
+
+  }
+
+  /**
+    * Represents the scheduler as a Sink and Source.
+    *
+    * This method will materialize the scheduler first, then Sink and Source can be materialized independently.
+    *
+    * @param specsSnapshot Snapshot of the current specs
+    * @return Snapshot of the current state, as well as Source which produces StateEvents and Sink which accepts SpecEvents
+    */
+  def asSourceAndSink(specsSnapshot: SpecsSnapshot): (Future[StateSnapshot], Source[StateEvent, Any], Sink[SpecUpdated, Any]) = {
```

Comment:
```suggestion
  def asSourceAndSink(specsSnapshot: SpecsSnapshot = SpecsSnapshot.empty): (Future[StateSnapshot], Source[StateEvent, Any], Sink[SpecUpdated, Any]) = {
```
Give it a default parameter and I don't believe we need the same method above